Added command line feature for remap_restarts.py

weiyuan-jiang opened the pull request:
This PR added command line features. The interface also changed. To get help, please use 
      ./remap_restarts.py -h
       ./remap_restarts.py config_file -h
       ./remap_restarts.py command_line -h

mathomp4 commented:
@weiyuan-jiang You should change to use the `choices` argument for argparser. For example, in my `create_expt.py` code I have:
```python
        p.add_argument('--moist', help="Moist microphysics to use (Default: GFDL)", type=str, default='BACM',
                choices=['BACM','GFDL','MGB2'])
```
and when I run `create_expt.py -h` I see:
```
  --moist {BACM,GFDL,MGB2}
                        Moist microphysics to use (Default: GFDL)
```

So argparse sort of automatically fills out the allowed choices which is nice *and* restricts it so you can't pass in unallowed options:
```
❯ create_expt.py test-c12 --moist ABCD
usage: create_expt.py [-h] [-v] [-q] [--expdsc EXPDSC] [--expdir EXPDIR]
...
create_expt.py: error: argument --moist: invalid choice: 'ABCD' (choose from 'BACM', 'GFDL', 'MGB2')
```

You can equivalently do something like:
```python
        moist_choices = ['BACM','GFDL','MGB2']
        p.add_argument('--moist', help="Moist microphysics to use (Default: GFDL)", type=str, default='BACM',
                choices=moist_choices)

```
I'm never quite sure when's the right time to make a separate array. I suppose if you can use the array in more than one place...

mathomp4 commented on pre/remap_restart/remap_bin2nc.py:
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python
```
@weiyuan-jiang Can you test this with:
```
#!/usr/bin/env python3
```
? I've been on a few systems that don't provide a `python` weirdly...or it points to `python2`.

mathomp4 commented on pre/remap_restart/remap_questions.py:
```diff
@@ -124,7 +128,11 @@ def ask_questions():
             "type": "text",
             "name": "output:shared:agrid",
-            "message": "Enter new atmospheric grid: \n C12   C180  C1000  C270\n C24   C360  C1440  C540\n C48   C500  C2880  C1080\n C90   C720  C5760  C2160  C1536\n ",
+            "message": "Enter new atmospheric grid: \n \
```
Just a question, but can `message` accept a triple-quoted string? Then maybe we wouldn't need the `\n`?

weiyuan-jiang replied:
Yes, it accepts. That is python's feature.

biljanaorescanin commented:
@gmao-rreichle  I've tests PR so far all tests I did worked. 
To note, at the moment this PR will not work for coupled grids until we update with new coupled grid name or we re-arrange original bcs sets they use. I wasn't sure should I pick their grids and reorganize to new structure? And which to take to do that.

gmao-rreichle commented on pre/remap_restart/remap_questions.py:
```diff
@@ -167,123 +157,88 @@ def ask_questions():
+    ICA: Icarus         5_17_0 ... 5_24_0_p1      Icarus, Jason    ...  10.18   
+    NL3: Icarus-NLv3    5_25_1 ... present        Icarus_NL, 10.19 ... present
+
+    Development:            Notes
+                        -------------------
+        v06:     NL3 + JPL veg height + PEATMAP + MODIS snow alb
```
I think these details should be moved to the "develop" query (~lines 209-212)?  At this point, the user only needs to know that there are additional "develop" options available.  If they select GM4, ICA, or NL3, they don't need to ever see the list of "develop" options.

gmao-rreichle commented on pre/remap_restart/remap_command_line.py:
```diff
@@ -0,0 +1,171 @@
+
+    p_command.add_argument('-ocnmdlin',   default='data',     help='Ocean model of input restarts',  choices=['data', 'MOM5', 'MOM6'])
+    p_command.add_argument('-ocnmdlout',  default='data',     help='Ocean model for new restarts', choices=['data', 'MOM5', 'MOM6'])
+    p_command.add_argument('-catch_model',default='catch',    help='Catchment[CN] model', choices=['catch', 'catchcnclm40', 'catchcnclm45'])
```
@weiyuan-jiang : Is this needed?  It's not in the "remap_questionary.py", so perhaps it's now deduced from the restart file name?

weiyuan-jiang replied:
I think we have to keep it here for command_line.  To make it simple, there is no deduction for command_line options

gmao-rreichle commented on pre/remap_restart/remap_command_line.py:
```diff
@@ -0,0 +1,171 @@
+    p_command.add_argument('-in_wemin',          help='Min. water snow water equivalent param. used with input restarts')
+    p_command.add_argument('-out_wemin',         help='Min. water snow water equivalent param. to be used with new restarts')
+
+    ocean_grids=['360x180','1440x720','2880x1440','CS', '72x36', '360x200','720x410','1440x1080']
```
@weiyuan-jiang : There is some duplication in the "choices" definitions between this script and remap_questions.py.  Can this be done in just one place and used in both scripts?  Also for `data model choices`, `coupled model choices`, and `['data', 'MOM5', 'MOM6']` below, and possibly other lists.

weiyuan-jiang replied:
Yes, it can be done

gmao-rreichle replied:
Thanks, @weiyuan-jiang, for avoiding the definition of the same lists multiple times **within** the `remap_command_line.py` script. This is better.
However, my original question went farther.  Can we use the lists defined in `remap_questions.py` (e.g., "choices_omodel", "choices_ogrid_data", "choices_ogrid_cpld", ...) to assemble the "choices" in `remap_command_line.py`?  That is, avoid hardwiring the same lists in two different scripts.
If this doesn't work, could we avoid defining lists in `remap_command_line.py` altogether and use "help" messages such as "see remap_questions.py for a list of choices"?

gmao-rreichle commented on pre/remap_restart/remap_restarts.py:
```diff
@@ -24,33 +24,33 @@
+           ./remap_restarts.py command_line -h
+
+      NOTE: Individual scripts can be executed independently:
+        1. remap_questions.py generates raw_answers.yaml.
+        2. If the command_line option is used with remap_restarts.py, the command line options are converted raw_answers.yaml.
+        3. remap_params.py uses raw_answer.yaml and remap_params.tpl as inputs and generates remap_params.yaml.
+        4. Sets of restarts can be remapped separately with remap_upper.py, remap_lake_landice_saltwater.py, 
+           remap_catchANDcn.py, and remap_analysis.py, all of which use remap_params.yaml as input for remapping.
```
I'm a bit confused about `raw_answers.yaml`, which is generated by `remap_questions.py` according to item 1.  I thought that `remap_questions.py` generates `remap_params.yaml`.  Can `remap_params.yaml` be hidden from the users in these instructions?

weiyuan-jiang replied:
raw_answers.yaml is created by remap_questions.py and was passed to remap_params.py to generate remap_params.yaml.  The raw_answers.yaml is hidden now. It is used when debugging.

gmao-rreichle replied:
I'm still a bit confused about the workings of the various scripts in this context, and how the "usage" text above can be edited without mentioning "raw_answers.yaml", given that "raw_answers.yaml" is hidden and used for debugging only.  If I understand correctly, currently `remap_questions.py` generates "raw_answers.yaml", which then makes it necessary to mention "raw_answers.yaml" in the "usage" text.  
Could we change `remap_questions.py` to generate "remap_params.yaml" by moving `params = remap_params(raw_config)` from `remap_restarts.py` to `remap_questions.py` (or maybe executing this in both scripts)?  
In any case, the "usage" text still needs editing, and I'm hoping it can be simplified by removing any mention of "raw_answers.yaml".

weiyuan-jiang replied:
The raw_answer.ymal was saved in a file before.  But it is not saved any more and the users won't know it. So we don't need to mention it any more.

gmao-rreichle commented on pre/remap_restart/remap_questions.py:
```diff
@@ -13,383 +13,374 @@
             "name": "input:shared:agrid",
-            "message": "Enter input atmospheric grid: \n C12   C180  C1000  C270\n C24   C360  C1440  C540\n C48   C500  C2880  C1080\n C90   C720  C5760  C2160  C1536\n ",
-            "default": 'C360',
+            "message": f'''Enter atmospheric grid of input restarts:
+    C12   C180  C1000  C270
+    C24   C360  C1440  C540
+    C48   C500  C2880  C1080
+    C90   C720  C5760  C2160  C1536\n''',
+            "validate": lambda text : text in ['C12','C180','C1000','C270','C24','C360','C1440','C540','C48','C500','C2880','C1080','C90','C720','C5760','C2160','C1536'],
```
Can we also put multi-line "message" strings into variables (like, e.g., `choices_bc_ops`)?  This particular block is needed twice, and it would be better to have it in just one place.  Likewise for the list in "validate".

weiyuan-jiang replied:
Yes, that makes the code cleaner

gmao-rreichle commented on pre/remap_restart/remap_questions.py:
```diff
@@ -13,383 +13,374 @@
-            "choices": ["discover_ops", "discover_legacy", "discover_couple","discover_ns", "other"],
-            "when": lambda x: not x['input:shared:MERRA-2'],
+            "name": "input:shared:bc_version",
+            "message": f'''\nSelect boundary conditions (BCs) version of input restarts:
+    BCs version      | ADAS tags            | GCM tags typically used with BCs version
+    -----------------|----------------------|-----------------------------------------
+    GM4: Ganymed-4_0 | 5_12_2 ... 5_16_5    | Ganymed-4_0      ... Heracles-5_4_p3
+    ICA: Icarus      | 5_17_0 ... 5_24_0_p1 | Icarus, Jason    ... 10.18   
+    NL3: Icarus-NLv3 | 5_25_1 ... present   | Icarus_NL, 10.19 ... present
+    ----------------------------------------------------------------------------------
+    Other: Additional choices used in model or DAS development.
+              \n\n ''',
```
Same comment as above.  Can this multi-line message (which occurs twice) be put into a variable?

gmao-rreichle commented on pre/remap_restart/remap_questions.py:
```diff
@@ -13,383 +13,374 @@
-            "when": lambda x: not x["input:shared:MERRA-2"] and not x["input:shared:bc_base"]== "discover_ns",
+            "type": "select",
+            "name": "input:shared:bc_version",
+            "message": f'''\nSelect BCs version of input restarts:
+
+             v06:     NL3 + JPL veg height + PEATMAP + MODIS snow alb\n\n''',
```
same comment as above - multi-line message that occurs twice

gmao-rreichle commented:
@weiyuan-jiang, I added [another commit](https://github.com/GEOS-ESM/GEOS_Util/pull/19/commits/4bc33f4f35a578cf91a89391b0629525c920ed19).  I don't think the centrally defined messages were correct.  I went out on a lib and tried my hand at concatenating multi-line strings and arrays of strings in Python. Chances are I messed up along the way, so don't assume that the code works.  I added the commit to illustrate what needs fixing.  I also revised the "usage" in remap_restarts.py.  Please proof-read.

gmao-rreichle commented on pre/remap_restart/remap_utils.py:
```diff
@@ -6,9 +6,258 @@
+message_bc_other_in  = ("Select BCs version of input restarts:\n" + message_bc_other)
+message_bc_other_new = ("Select BCs version for new restarts:\n"  + message_bc_other)
+
+message_agrid_list = f'''
+ C12   C180  C540   C1440
+ C24   C270  C720   C1536
+ C48   C360  C1000  C2160
+ C90   C500  C1080  C2880  C5760\n'''
+
+message_agrid_in   = ("Enter atmospheric grid of input restarts:\n" + message_agrid_list)
+
+message_agrid_new  = ("Enter atmospheric grid for new restarts:\n"  + message_agrid_list)
+
+validate_agrid     = ['C12','C24','C48','C90','C180','C270','C360','C500','C540','C720','C1000','C1080','C1440','C1536','C2160','C2880','C5760']
```
@weiyuan-jiang, @biljanaorescanin : The two lists of resolutions were in no sensible order.  I reordered them.  Do we really use all these resolutions?  @biljanaorescanin, please double-check against the bcs that are available and check if any of the resolutions can possibly be deleted.  We could leave this to when we address the Stretched_CS changes

weiyuan-jiang commented:
> @weiyuan-jiang, I added [another commit](https://github.com/GEOS-ESM/GEOS_Util/pull/19/commits/4bc33f4f35a578cf91a89391b0629525c920ed19). I don't think the centrally defined messages were correct. I went out on a lib and tried my hand at concatenating multi-line strings and arrays of strings in Python. Chances are I messed up along the way, so don't assume that the code works. I added the commit to illustrate what needs fixing. I also revised the "usage" in remap_restarts.py. Please proof-read.

Looks good to me. Thanks for fixing it.

gmao-rreichle commented:
@biljanaorescanin:   This PR is nearly ready.  I think we should wrap it up now and then start a new PR that adds functionality for the coupled atm/ocean grids.  To wrap up the PR, two things are needed:
1) Address the following comment: https://github.com/GEOS-ESM/GEOS_Util/pull/19/files#r1327818173
2) Verify that the remapping still works after the latest commits.

I don't think "nightly" GCM or LDAS tests are needed because the PR should be trivially 0-diff for GCM or LDAS runs, but I'll certainly defer to Scott on this  when we hand the PR over to him. 

cc: @weiyuan-jiang

gmao-rreichle commented on pre/remap_restart/remap_upper.py:
```diff
@@ -131,35 +136,37 @@ def remap(self):
+     stretch = config['output']['shared']['stretch']
+     stretch_str = ""
+     if stretch:
+        target_lat  = 0.0
+        target_lon  = 0.0
+        stretch_fac = 1.0
+        if stretch == 'SG001':
```
1. I don't think we should set these defaults and then leave them in place if the particular string for SGxxx is not in list of "if" statements.  If for some reason (e.g., incomplete code update), we arrive here with stretch='SG003', then the parameters will be the defaults, right?  But the defaults are by definition the not-stretched CS grid.  We need to make sure that the code exits cleanly if "stretch" is unknown.
2. More broadly, I think we need to define the SGxxx table somewhere centrally, say in remap_utils.py.  At the moment, there are two places where SG001 and SG002 are defined, not incl. the comments.

gmao-rreichle replied:
Mostly addressed by the two most recent commits.  Could still use more cleanup.

gmao-rreichle commented on pre/remap_restart/remap_upper.py:
```diff
@@ -131,35 +136,37 @@ def remap(self):
-
      # If we are running stretched grid, we need to pass in the target lat, lon, and stretch factor
      # to interp_restarts.x. Per the code we use:
      #  -stretched_grid target_lon target_lat stretch_fac
```
This is a bit nitpicky, but I think it would be best to use the same order of these three parameters everywhere.  In the comments and the sg001 and sg002 defs in remap_utils.py, we use the order fac/lat/lon.  In the lines below, we use lat/lon/fac, and in the cmd line arg list we use lon/lat/fac.  It's easy to see how someone will accidentally get the order wrong.

gmao-rreichle replied:
Mostly addressed in most recent commit https://github.com/GEOS-ESM/GEOS_Util/pull/19/commits/40e43d66028852f696b8fa26189e6dd5ff36af51

gmao-rreichle commented on pre/remap_restart/remap_utils.py:
```diff
@@ -6,9 +6,237 @@
+  x['input:shared:stretch'] = False
+  stretch_factor = fvrst.__dict__.get('STRETCH_FACTOR')
+  if (stretch_factor) :
+     sg001 = ['2.50', '39.50', '-98.35']
+     sg002 = ['3.00', '39.50', '-98.35']
```
This should be defined centrally.  See comment in remap_upper.py.

gmao-rreichle commented on pre/remap_restart/remap_utils.py:
```diff
@@ -6,9 +6,237 @@
+     sg = [stretch_factor, target_lat, target_lon]
+     f_ = [f"{number:.{2}f}" for number in sg]
+     print("[stretch_factor, target_lat, target_lon]: ", f_)
+     if f_ == sg001 :
+       x['input:shared:stretch'] = 'SG001'
+     if f_ == sg002 :
+       x['input:shared:stretch'] = 'SG002'
```
Need to exit cleanly if "stretch" is not in the list of known names.

gmao-rreichle commented:
@biljanaorescanin : I added another commit that further cleans up the handling of the stretched CS parameters.  This isn't perfect, but it's good enough for now.  I probably got the python syntax wrong, so chances are it won't work out of the box - but it just may.  Can you please take another look?  If we can get this to work and produce the 0-diff restarts vs. the old regrid.pl, we could move it onto Scott's desk.

gmao-rreichle commented on pre/remap_restart/remap_questions.py:
```diff
@@ -13,383 +13,389 @@
+  return False
+
+def echo_level(x):
+  if x["output:air:nlevel"] != "72":
```
@biljanaorescanin, @weiyuan-jiang: Is this if condition general enough? If I'm reading it correctly, the assumption here is that the number of atm levels in the **input** restarts is always 72.  What if somebody wants to remap from, say, 131 to 181 levels?   Can we just replace the "72" with "input:air:nlevel"??   (Also in line 42.)  Not sure if I understand this correctly.

biljanaorescanin replied:
With last commit @weiyuan-jiang fixed this so its more flexible.

biljanaorescanin commented:
I've tested PR for my usual cases, and all are zero diff from what we had before. 
I've also tested new functionalities (stretched grids, new BCS versions), and those also work. 

@sdrabenh I think this is ready for you to take a look.  If all pass your testing, we can merge this PR as it is now and continue to work on new PR to add coupled grids since now the MOM5/MOM6 option is just a placeholder. (But all other functionalities should work. )

tclune reviewed (dismissed):
cmake changes ok

sanAkel reviewed (changes requested):
@weiyuan-jiang can you add a notebook (clear output before pushing to github) to show at least one use case, for e.g., `remap_upper.py`? Thanks!

gmao-rreichle commented:
> @weiyuan-jiang can you add a notebook (clear output before pushing to github) to show at least one use case, for e.g., `remap_upper.py`? Thanks!

@sanAkel - we are catching up with the documentation.  Our plan is to update the [Wiki page](https://github.com/GEOS-ESM/GEOS_Util/wiki).  Please bear with us.

weiyuan-jiang commented:
> @weiyuan-jiang can you add a notebook (clear output before pushing to github) to show at least one use case, for e.g., `remap_upper.py`? Thanks!

I have modified the wiki page. That would be our starting point for now.  Please help us improve the wiki by commenting on it @sanAkel @JulesKouatchou @gmao-rreichle

weiyuan-jiang commented:
The last commit resolves the issue #37

biljanaorescanin commented:
I've tested my cases and the automatic ones we offer as an option with the PR, and all passed zero diff. 

Note: We will need to update the baseline stored at `*/projects/gmao/SIteam/Remapping_Test_Cases/` since these test runs we use to compare, ( baseline) from regrid.pl are outdated and not correct for upper restarts. 

Coupled cases also work, but only for resolutions we have boundary conditions for. (that is the same as before, but just a note for people who don't know and receive email)

gmao-rreichle commented on pre/remap_restart/remap_utils.py:
```diff
@@ -6,9 +6,268 @@
+# define "bc_base", "choices", "message" strings, and "validate" lists that are used multiple times
+#   (and related definitions, even if they are used just once).
+
+bc_base = "/discover/nobackup/projects/gmao/bcs_shared/fvInput/ExtData/esm/tiles"
```
@biljanaorescanin, @weiyuan-jiang :  `bc_base` here is hardwired for Discover.  Does this mean that the tool won't work at NAS?  I didn't see an opportunity for users to overwrite `bc_base` (but might have missed it).
I suppose we will need to keep a copy of the official bcs dir at NAS if we're asking users to run off the official bcs (in progress PRs for gcm_setup/fv_setup ).    Then we may as well point remap_restarts to the corresponding NAS directory.  It's up to Scott to decide if this functionality has to be present before he agrees to merge this PR.  I'm ok either way, but we'll definitely need to work on this sooner or later.

mathomp4 replied:
It does not! I don't think there is any reason it couldn't work at NAS if we had all the relevant BCs there (which we probably do though not the new BCs yet).

We'd also of course have to fix up all the bits about "what node type" do you want to run and sbatch v qsub, etc.

weiyuan-jiang replied:
So far it won't be working on NAS. But After NAS is ready for the new bcs, I will change this [line ](https://github.com/GEOS-ESM/GEOS_Util/blob/e84dddee8fbad8b263cfbd1169893d82ffb8b670/pre/remap_restart/remap_utils.py#L481)

@gmao-rreichle

gmao-rreichle replied:
@weiyuan-jiang, thanks for the changes.  Once we know bc_base for NAS we can add it.  At that point, the tool will work at NCCS and NAS.  However, we still need the functionality for users to supply their own bc_base value.  Not all users are at NCCS or NAS.  Such users would have to edit `remap_utils.py` to get the tool to work for them, which isn't something we should ask of them.  The solution would be to include a question for bc_base in the questionary.  We can detect if we're at NCCS or NAS, and in this case we can skip the question and use the paths that are hardwired into `remap_utils.py`.  But if we're not at NCCS or NAS, we need to ask the user for bc_base.  
Since this functionality is needed anyway, we should probably do it now, because it will also make the code work at NAS immediately.  A NAS user just has to copy the (revised layout) bcs that they're using from NCCS into their own NAS directory temporarily.  Not great, but I'm still scarred from having to wait for the NCCS bcs_shared project directory for 12+ months.

weiyuan-jiang replied:
The question for bc_base is added. Also a 'customized' option is added for users' own base,

gmao-rreichle commented on pre/remap_restart/tests/f522Toc360.yaml:
```diff
@@ -45,4 +52,4 @@ output:
 slurm:
   account: g0610
```
The yaml files of the five test cases all have an NCCS account (sponsor code) hardwired, although the account number differs between yaml files.  Is the account entry needed in the yaml files for the test cases?  Doesn't the test_remap_restarts.py script determine the account number?

weiyuan-jiang replied:
set to blank

gmao-rreichle commented on pre/remap_restart/remap_params.tpl:
```diff
@@ -8,11 +8,16 @@ input:
     bcs_dir:
     expid:
     ogrid:
+    # (coupled) ocean model: data, MOM5, MOM6
+    omodel: data
```
It probably doesn't matter to python, but the order of the params in this template differs from that in the yaml files for the test cases.  To get a clean start with the remap_restarts.py, it may make sense to re-issue the yaml files for the test cases so they conform with the latest template file before merging the PR into develop.

weiyuan-jiang replied:
re-ordered

gmao-rreichle commented on pre/remap_restart/remap_questions.py:
```diff
@@ -1,395 +1,448 @@
             "type": "confirm",
             "name": "input:shared:MERRA-2",
-            "message": "Would you like to remap archived MERRA-2 restarts?",
+            "message": "Remap archived MERRA-2 restarts? (NCCS/Discover only; elsewhere, select 'N' and complete full config; requires nc4 restarts.)\n",
```
@weiyuan-jiang, @biljanaorescanin : I assume the option to use archived M-2 restarts works on Discover only and clarified the question accordingly.  I may be wrong.  Scott and/or Matt can advise on whether this is something that should work at NAS, in which case more work is needed.  I doubt this is urgent, though, and it can be left for later.

sdrabenh replied:
Most likely M2 restarts would have to be migrated to NAS. However, that is something which could be addressed later if there is a need to do so.

gmao-rreichle commented on pre/remap_restart/remap_utils.py:
```diff
@@ -1,14 +1,286 @@
+
+# top-level directory for BCs (machine-dependent)
+
+choices_bc_base  =[ "NCCS/Discover : /discover/nobackup/projects/gmao/bcs_shared/fvInput/ExtData/esm/tiles",
```
@weiyuan-jiang : Thanks for adding the question about bc_base.  However, don't we want to do this a bit differently?  That is, first detect if we are on Discover or at NAS, and automatically select the hardwired bc_base dir accordingly, without bothering the user.  The question is needed only when the user is not at NCCS or NAS.  I suppose we can fix this later, as it does not impact the yaml interface and only touches on the user experience.  But then again, it might be easier to fix it now while we're at it.

weiyuan-jiang replied:
I think this provide the users an opportunity  to specify their own bc_base.

gmao-rreichle commented:
> @weiyuan-jiang can you add a notebook (clear output before pushing to github) to show at least one use case, for e.g., `remap_upper.py`? Thanks!

@sanAkel -- we have updated the Wiki page: https://github.com/GEOS-ESM/GEOS_Util/wiki/remap_restarts-package  and ironed out a few more kinks (incl. making the package ready for NAS).   Please take another look and see if this works for you.   If so, please approve, so we can put this on @sdrabenh's desk.   I'll defer to @sdrabenh on whether a notebook is needed, and we can always add it later.  My preference would be to move this forward before a potential shutdown.  Note also that we provide pre-configured test cases / examples in the ./test dir.    Thanks! 

cc: @weiyuan-jiang @biljanaorescanin

gmao-rreichle reviewed (approved):
re-approving after bug fix

biljanaorescanin commented:
@weiyuan-jiang we will need few more changes to run this on NAS. We did rsync BCS files but seems they use `PBS` to submit jobs there and now our script is `SBATCH` hardcoded. 

For everything else this PR was tested:
1.   automatic tests that are part of package ( this will work only on discover since NAS doesn't have baseline tests for comparison. )
2. 12 more random cases which also included few coupled cases where we just change resolution. 

I couldn't find any issues.  @sdrabenh let me know if you find something during your testing.

sdrabenh commented:
@mathomp4 can you approve on behalf of cmake-team? @bena-nasa doesn't appear to have permissions

mathomp4 commented:
> @mathomp4 can you approve on behalf of cmake-team? @bena-nasa doesn't appear to have permissions

Odd. He definitely should! I'll look into that next week

sanAkel commented:
> > @weiyuan-jiang can you add a notebook (clear output before pushing to github) to show at least one use case, for e.g., `remap_upper.py`? Thanks!
> 
> 
> 
> @sanAkel -- we have updated the Wiki page: https://github.com/GEOS-ESM/GEOS_Util/wiki/remap_restarts-package  and ironed out a few more kinks (incl. making the package ready for NAS).   Please take another look and see if this works for you.   If so, please approve, so we can put this on @sdrabenh's desk.   I'll defer to @sdrabenh on whether a notebook is needed, and we can always add it later.  My preference would be to move this forward before a potential shutdown.  Note also that we provide pre-configured test cases / examples in the ./test dir.    Thanks! 
> 
> 
> 
> cc: @weiyuan-jiang @biljanaorescanin 
> 
> 

Thanks @gmao-rreichle wiki looks self-sufficient.
